File: unique-binary-search-trees-ii.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; left = null; right = null; }
 * }
 */
public class Solution {
    public ArrayList<TreeNode> generateTrees(int n) {
        return generate(1, n);
    }
    
    private ArrayList<TreeNode> generate(int p, int q) {
        ArrayList<TreeNode> ans = new ArrayList<TreeNode>();
        if (p > q) {
            ans.add(null);
            return ans;
        } else if (p == q) {
            ans.add(new TreeNode(p));
            return ans;
        }
        for (int i = p; i <= q; i++) {
            ArrayList<TreeNode> lefts = generate(p, i - 1);
            ArrayList<TreeNode> rights = generate(i + 1, q);
            for (TreeNode left : lefts) {
                for (TreeNode right : rights) {
                    TreeNode node = new TreeNode(i);
                    node.left = left;
                    node.right = right;
                    ans.add(node);
                }
            }
        }
        return ans;
    }
}

File: unique-binary-search-trees-ii.cc
#include <cstdlib>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  vector<TreeNode*> generateTrees(int n) {
    return generate_trees(1, n);
  }

  vector<TreeNode*> generate_trees(int l, int r) {
    vector<TreeNode*> ret;
    if (l > r) {
      ret.push_back(NULL);
    } else if (l == r) {
      ret.push_back(new TreeNode(l));
    } else {
      for (int i = l; i <= r; i++) {
        vector<TreeNode*> left_trees = generate_trees(l, i - 1);
        vector<TreeNode*> right_trees = generate_trees(i + 1, r);
        for (int j = 0; j < left_trees.size(); j++) {
          for (int k = 0; k < right_trees.size(); k++) {
            TreeNode* root = new TreeNode(i);
            root->left = left_trees[j];
            root->right = right_trees[k];
            ret.push_back(root);
          }
        }
      }
    }
    return ret;
  }
};
